middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml

```diff
@@ -814,74 +826,114 @@ let bigarray_set ~dbg ~unsafe kind layout b indexes value =
+    | Array_kind Immediates
+    | Array_kind Values
+    | Array_kind Naked_floats
+    | Array_kind Naked_float32s
+    | Array_kind Naked_int32s
+    | Array_kind Naked_int64s
+    | Array_kind Naked_nativeints
+    | Float_array_opt_dynamic ->
+      1
+  in
   checked_access ~primitive
     ~conditions:
       (array_access_validity_condition array array_kind index ~index_kind
-         ~size_int)
+         ~width_in_scalars ~size_int)
     ~dbg
 
-let array_load_unsafe ~array ~index (array_ref_kind : Array_ref_kind.t)
-    ~current_region : H.expr_primitive =
+let array_load_unsafe ~array ~index _array_kind
+    (array_ref_kind : Array_ref_kind.t) ~current_region : H.expr_primitive list
+    =
   match array_ref_kind with
-  | Immediates -> Binary (Array_load (Immediates, Scalar, Mutable), array, index)
-  | Values -> Binary (Array_load (Values, Scalar, Mutable), array, index)
+  | Immediates ->
+    [Binary (Array_load (Immediates, Immediates, Mutable), array, index)]
+  | Values -> [Binary (Array_load (Values, Values, Mutable), array, index)]
   | Naked_floats_to_be_boxed mode ->
-    box_float mode
-      (Binary (Array_load (Naked_floats, Scalar, Mutable), array, index))
-      ~current_region
+    [ box_float mode
+        (Binary (Array_load (Naked_floats, Naked_floats, Mutable), array, index))
+        ~current_region ]
   | Naked_floats ->
-    Binary (Array_load (Naked_floats, Scalar, Mutable), array, index)
+    [Binary (Array_load (Naked_floats, Naked_floats, Mutable), array, index)]
   | Naked_float32s ->
-    Binary (Array_load (Naked_float32s, Scalar, Mutable), array, index)
+    [Binary (Array_load (Naked_float32s, Naked_float32s, Mutable), array, index)]
   | Naked_int32s ->
-    Binary (Array_load (Naked_int32s, Scalar, Mutable), array, index)
+    [Binary (Array_load (Naked_int32s, Naked_int32s, Mutable), array, index)]
   | Naked_int64s ->
-    Binary (Array_load (Naked_int64s, Scalar, Mutable), array, index)
+    [Binary (Array_load (Naked_int64s, Naked_int64s, Mutable), array, index)]
   | Naked_nativeints ->
-    Binary (Array_load (Naked_nativeints, Scalar, Mutable), array, index)
-
-let array_set_unsafe ~array ~index ~new_value
-    (array_set_kind : Array_set_kind.t) : H.expr_primitive =
-  let new_value =
-    match array_set_kind with
-    | Immediates | Values _ | Naked_floats | Naked_float32s | Naked_int32s
-    | Naked_int64s | Naked_nativeints ->
-      new_value
-    | Naked_floats_to_be_unboxed -> unbox_float new_value
+    [ Binary
+        (Array_load (Naked_nativeints, Naked_nativeints, Mutable), array, index)
+    ]
+
+let array_load_unsafe ~array ~index array_kind array_load_kind ~current_region =
+  array_load_unsafe ~array ~index array_kind array_load_kind ~current_region
+  |> H.maybe_create_unboxed_product
+
+let array_set_unsafe dbg ~array ~index array_kind
+    (array_set_kind : Array_set_kind.t) ~new_values : H.expr_primitive list =
+  let[@inline] normal_case array_set_kind new_values =
```

Comment:
Not sure the `[@inline]` is useful here, since this is a local function so this should probably be `[@local]`?

Reply:
I've tended to use `[@inline]` as sometimes `[@local]` is refused, but it does work in this case, so fine.